### PR TITLE
Event Handling Infrastructure

### PR DESCRIPTION
**Pull request by codeNinjaDev**
Basic code to manage handling of incoming events via event reader.

Jira: https://cdap.atlassian.net/browse/CDAP-20678
Interacting with [event-reader-spi](https://github.com/cdapio/cdap/tree/develop/cdap-event-reader-spi/src/main/java/io/cdap/cdap/spi/events)

The **EventSubscriberManager** will initialize all EventSubscribers (e.g StartProgramEventSubscriber, etc...) and start them if the event reading feature is enabled.

Each **EventSubscriber** (a ScheduledService) will handle a specific type of event and will initialize its respective EventReader with the DefaultEventReaderContext.

The **EventReaderExtensionProvider** provides the available EventReaders to the EventSubscribers.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+   *
+   * @param cConf An instance of an injected ${@link CConfiguration}.
+   * @param id class name of the event reader extension. E.g.: pub_sub_event_readerA
+   */
+  DefaultEventReaderContext(CConfiguration cConf, String id) {
+    String prefix = String.format("%s.%s.", Constants.Event.EVENTS_READER_PREFIX, id);
+
+    int publishTimeout = cConf.getInt(Constants.AppFabric.PROGRAM_STATUS_RETRY_STRATEGY_PREFIX
+            + Constants.Retry.MAX_TIME_SECS);
+
```
Comment (codeNinjaDev) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java]: Is this really worth it? If buffer > 0 on default publish timeout, it will automatically fail (max ack deadline == default publish timeout).

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java [hunk 2]
```diff
@@ -221,51 +219,25 @@ public final ProgramController run(final Program program, ProgramOptions oldOpti
...
       // For runs from a tethered instance, copy over peer runtime token if it exists.
       if (oldOptions.getArguments().hasOption(ProgramOptionConstants.PEER_NAME)) {
         copyRuntimeToken(locationFactory, oldOptions.getArguments(), tempDir, localizeResources);
       }
 
-      Map<String, String> extraSystemArgs = new HashMap<>(launchConfig.getExtraSystemArguments());
-
       // Localize the program jar
       Location programJarLocation = program.getJarLocation();
-      String programJarName = programJarLocation.getName();
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java]: What is the context for these changes ?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventHandler.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package io.cdap.cdap.internal.events;
+
+import com.google.common.util.concurrent.AbstractScheduledService;
+
+/**
+ * Abstract class for an event Handler. Each EventHandler will receive and process incoming events
+ */
+public abstract class EventHandler extends AbstractScheduledService {
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventHandler.java]: EventHandler is a very generic name. Suggest using a more specific one.

File: cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java [hunk 4]
```diff
@@ -378,6 +384,11 @@ String getGcsBucket() {
     return gcsBucket;
   }
 
+  @Nullable
+  String getTempBucket() {
```
Comment (greeshmaswaminathan) [inline cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Are these changes required for event handling?

Comment (codeNinjaDev) [reply cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocConf.java]: Good catch, these are not required. Somehow got added to commit

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventHandler.java (hunk 3 above)
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventHandler.java]: How does `IncomingEventHandler` sound?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java [hunk 5]
```diff
@@ -259,7 +259,7 @@ public final ProgramController run(final Program program, ProgramOptions oldOpti
         // Also extra files to the resources jar so that the TwillAppLifecycleEventHandler,
         // which runs in the AM container, can get them.
         // This can be removed when TWILL-246 is fixed.
-        // Only program running in Hadoop will be using EventHandler
+        // Only program running in Hadoop will be using EventReaderHandler
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java]: Why this change?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  private final Map<String, String> properties;
+
+  /**
+   * Construct the default Event reader context.
+   *
+   * @param eventTypePrefix type of event to handle: eg, start, deploy, etc...
+   * @param className    simple class name of event reader
+   * @param cConf        An instance of an injected ${@link CConfiguration}.
+   */
+  DefaultEventReaderContext(String eventTypePrefix, String className, CConfiguration cConf) {
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java]: Let's pass className as part of the prefix since we may change it in the future. Constructor should only have two arguments: prefix and cConf

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java (hunk 1 above)
Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java]: Not following. Can you elaborate.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventReaderHandlerManager.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    this.eventReaderHandlers = eventReaderHandlers;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    if (!enabled) {
+      return; // If not enabled, don't start
+    }
+    eventReaderHandlers.forEach(eventReaderHandler -> {
+      if (eventReaderHandler.initialize()) {
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventReaderHandlerManager.java]: Would be good to log handlers' initialize outcomes.

File: cdap-features/src/main/java/io/cdap/cdap/features/Feature.java [hunk 8]
```diff
@@ -28,6 +28,7 @@
 public enum Feature {
   REPLICATION_TRANSFORMATIONS("6.6.0"),
   EVENT_PUBLISH("6.7.0", false),
+  EVENT_READ("6.10.0", false),
```
Comment (masoud-io) [inline cdap-features/src/main/java/io/cdap/cdap/features/Feature.java]: nit: EVENT_READER

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java (hunk 5 above)
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java]: Intelli-j refactor, will fix

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java (hunk 1 above)
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java]: Added some comments to the code to clarify. Basically, the maximum acknowledgement deadline is 10 minutes. The default maximum retry timeout is also 10 minutes. If the retry timeout is shorter than the acknowledgement deadline, the likelihood of duplicate runs (programs published to TMS) due to being incorrectly redelivered is reduced. However, this will likely involve decreasing the default retry timeout which may be a dependency for other CDAP functionality.

File: cdap-app-fabric/pom.xml [hunk 9]
```diff
@@ -269,6 +269,12 @@
       <artifactId>cdap-source-control</artifactId>
       <version>${project.version}</version>
     </dependency>
+      <dependency>
+          <groupId>io.cdap.cdap</groupId>
+          <artifactId>cdap-event-reader-spi</artifactId>
+          <version>${project.version}</version>
+          <scope>compile</scope>
```
Comment (masoud-io) [inline cdap-app-fabric/pom.xml]: why do we need scope?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  private final Map<String, String> properties;
+
+  /**
+   * Construct the default Event reader context.
+   *
+   * @param prefix prefix for specific event reader (e.g start.simpleclassname)
+   * @param cConf        An instance of an injected ${@link CConfiguration}.
+   */
+  DefaultEventReaderContext(String prefix, CConfiguration cConf) {
+    String completePrefix = String.format("%s.%s", Constants.Event.EVENTS_READER_PREFIX, prefix);
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java]: Do you expect to use EVENTS_READER_PREFIX in any other place besides here? If no, it'd be better to define it as a final constant in this class.

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java [hunk 11]
```diff
@@ -2219,6 +2219,10 @@ public static final class Event {
     public static final String EVENTS_WRITER_EXTENSIONS_DIR = "events.writer.extensions.dir";
 
     public static final String EVENTS_WRITER_EXTENSIONS_ENABLED_LIST = "events.writer.extensions.enabled.list";
+
+    public static final String EVENTS_READER_PREFIX = "event.reader";
+
+    public static final String EVENTS_READER_ACK_BUFFER = "events.reader.buffer";
```
Comment (masoud-io) [inline cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java]: You need to set this in cdap-default.xml.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+  private final Map<String, String> properties;
+
+  /**
+   * Construct the default Event reader context.
+   *
+   * @param prefix prefix for specific event reader (e.g start.simpleclassname)
+   * @param cConf        An instance of an injected ${@link CConfiguration}.
+   */
+  DefaultEventReaderContext(String prefix, CConfiguration cConf) {
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java:35]: What if some required context (e.g., topic) are not provided? We may wanna have a validate method to validate that all required context has been set.

Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java:35]: It's easier to validate in the extension than this class, especially since different readers may need require different context.

Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java:35]: Sounds good. Let's make sure to add validation and throws proper exception with explanation if full context is not provided.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java (hunk 10 above)
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java]: Rather than combining `Constants.Event.EVENTS_READER_PREFIX` and the `prefix` argument here, what do you think of combining them before passing it into the constructor at the caller-side?

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java (hunk 11 above)
Comment (codeNinjaDev) [reply cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java]: Both constants?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * Construct the default Event reader context.
+   *
+   * @param prefix prefix for specific event reader
+   * @param cConf        An instance of an injected ${@link CConfiguration}.
+   */
+  DefaultEventReaderContext(String prefix, CConfiguration cConf) {
+
+    Map<String, String> mutableProperties = new HashMap<>(cConf.getPropsWithPrefix(prefix));
+    mutableProperties.put("ackDeadline", cConf.get(Constants.Event.EVENT_READER_ACK_DEADLINE,
```
Comment (codeNinjaDev) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java]: rather than having ack deadline identical for all event readers, maybe enforce that it is specified on the extension-side as one of the entries loaded by `cConf.getPropsWithPrefix`?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+   * Construct the default Event reader context.
+   *
+   * @param prefix prefix for specific event reader
+   * @param cConf        An instance of an injected ${@link CConfiguration}.
+   */
+  DefaultEventReaderContext(String prefix, CConfiguration cConf) {
+    Map<String, String> mutableProperties = new HashMap<>(cConf.getPropsWithPrefix(prefix));
+    mutableProperties.put("ackDeadline", cConf.get(Constants.Event.EVENT_READER_ACK_DEADLINE,
+            String.valueOf(DEFAULT_ACK_DEADLINE)));
+    this.properties = Collections.unmodifiableMap(mutableProperties);
```
Comment (codeNinjaDev) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java:38]: Need to also add instance name to properties

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/dummy/DummyEventReaderExtensionProvider.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+import io.cdap.cdap.spi.events.Event;
+import io.cdap.cdap.spi.events.EventReader;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Dummy implementation for {@link EventReaderProvider} for test proposals.
+ */
+public class DummyEventReaderExtensionProvider<T extends Event> implements EventReaderProvider {
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/dummy/DummyEventReaderExtensionProvider.java]: Are these dummy implementations used right now ?

Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/dummy/DummyEventReaderExtensionProvider.java]: It's used in the StartProgramEventSubscriberTest as a dependency for StartProgramEventSubscriber

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+ * Each EventSubscriber will receive and process incoming events.
+ */
+public abstract class EventSubscriber extends AbstractScheduledService {
+
+  /**
+   * Initialize this handler.
+   *
+   * @return true if successful
+   */
+  public abstract boolean initialize();
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java]: For simplicity, I'm thinking of making this method void, and catch/log if it throws an exception in EventSubscriberManager. Otherwise, we can treat it as successfully initialized.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+  @Override
+  protected void startUp() throws Exception {
+    if (!enabled) {
+      return; // If not enabled, don't start
+    }
+    eventSubscribers.forEach(eventSubscriber -> {
+      // Loading the event readers from provider
+      // Initialize the event subscribers with all the event readers provided by provider
+      if (eventSubscriber.initialize()) {
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java]: an initialize method may throw an exception. We need to catch the exception.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+        eventSubscriber.startAndWait();
+        LOG.info("Successfully initialized eventSubscriber: {}",
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java:54]: nit: Do we have access to its reader id so we can include it in the log?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java (hunk 16 above)
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriber.java]: What exception should it throw? If it can't load any event readers? In the StartProgramEventSubscriber, if one event reader can't load, it doesn't impact the other ones: https://github.com/cdapio/cdap/blob/e87b96137107f3f22283211f4c7446c21f1a45a7/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java#L83

If no readers were able to be loaded, should I  throw an exception and handle it in the manager?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java (hunk 18 above)
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java:54]: There can be multiple event readers per event subscriber.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+  /**
+   * Construct the default Event reader context.
+   *
+   * @param prefix prefix for specific event reader
+   * @param cConf        An instance of an injected ${@link CConfiguration}.
+   */
+  DefaultEventReaderContext(String prefix, CConfiguration cConf) {
+    Map<String, String> mutableProperties = new HashMap<>(cConf.getPropsWithPrefix(prefix));
+    mutableProperties.put(Constants.INSTANCE_NAME, cConf.get(Constants.INSTANCE_NAME));
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java:37]: What is instance name used for? Is this available in OSS deployment ?

File: cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java
Comment (greeshmaswaminathan) [inline cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java:1]: Seems like no changes in this file?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java (hunk 19 above)
Comment (codeNinjaDev) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/DefaultEventReaderContext.java:37]: When publishing failure, the extension will publish the exception as well as which instance it failed on

**Comment by codeNinjaDev** [comment]
Tested in sandbox (without enabling event subscribing), deploying and running a pipeline is successful

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    });
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (!enabled) {
+      return; // If not enabled, don't shut down
+    }
+    eventSubscribers.forEach(eventSubscriber -> {
+      eventSubscriber.stopAndWait();
```
Comment (masoud-io) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java]: We should log if it throws an exception.

**Review by masoud-io — COMMENTED** [review]
One comment. LGTM otherwise.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+    this.eventSubscribers = eventSubscribers;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    if (!enabled) {
+      return; // If not enabled, don't start
+    }
+    eventSubscribers.forEach(eventSubscriber -> {
+      // Loading the event readers from provider
```
Comment (greeshmaswaminathan) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java]: The comments looks outdated here.

**Review by greeshmaswaminathan — APPROVED** [review]

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    if (!enabled) {
+      return; // If not enabled, don't start
+    }
+    eventSubscribers.forEach(eventSubscriber -> {
+      // Initialize the event subscribers with all the event readers provided by provider
+      try {
+        eventSubscriber.initialize();
+        eventSubscriber.startAndWait();
+        LOG.info("Successfully initialized eventSubscriber: {}",
+            eventSubscriber.getClass().getSimpleName());
```
Comment (codeNinjaDev) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java:55]: @masoud-io 
Should I standardize the way I log the eventSubscriber? (simplename vs classname vs other)

Comment (masoud-io) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventSubscriberManager.java:55]: Yes, that'd be great. You can do it in a separate PR also.

**Review by masoud-io — APPROVED** [review]
